AX: REGRESSION(286869@main): AXPropertyName::IsExpanded becomes stale for details and summary elements, causing accessibility/mac/details-summary.html to fail in ITM

twilco opened the pull request:
#### 9da9cefb7d6d75043aef967725b987292d3863f2
<pre>
AX: REGRESSION(286869@main): AXPropertyName::IsExpanded becomes stale for details and summary elements, causing accessibility/mac/details-summary.html to fail in ITM
<a href="https://bugs.webkit.org/show_bug.cgi?id=283578">https://bugs.webkit.org/show_bug.cgi?id=283578</a>
<a href="https://rdar.apple.com/140428627">rdar://140428627</a>

Reviewed by Chris Fleizach.

Prior to <a href="https://commits.webkit.org/286869@main">https://commits.webkit.org/286869@main</a>, we used to get away with not updating this property because every time
the details were expanded, we would throw away the entire details, summary, and other descendant objects and create new
ones (because the DOM implementation destroyed and recreated the DOM elements each time). After
<a href="https://commits.webkit.org/286869@main">https://commits.webkit.org/286869@main</a>, the DOM implementation no longer destroys and recreates brand new DOM elements,
instead managing the visibility via CSS content-visibility. This a good change, since we now do significantly less work,
but need to make sure this property stays up-to-date. This commit does that.

* LayoutTests/accessibility/mac/details-summary-expected.txt:
* LayoutTests/accessibility/mac/details-summary.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onExpandedChanged):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::toggleOpen):

Canonical link: <a href="https://commits.webkit.org/287009@main">https://commits.webkit.org/287009@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad083bdb537d0c284003ebc2d5af4392dad1ac98

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77906 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31281 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82554 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29163 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5234 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61002 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18935 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66907 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41305 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48359 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24365 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27506 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69453 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24718 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83916 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5273 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3549 "Found 2 new test failures: imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic.html ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69222 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5429 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66805 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68478 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17108 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12476 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10592 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5221 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7974 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5213 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8645 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6998 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1313,6 +1315,17 @@ void AXObjectCache::onEventListenerRemoved(Node& node, const AtomString& eventTy
 
+void AXObjectCache::onExpandedChanged(HTMLDetailsElement& detailsElement)
+{
+    postNotification(get(detailsElement), AXExpandedChanged);
```
do we want to post this after ITM is scheduled to do work? It seems possible that there could be a gap where the notification is posted, but ITM is scheduled to update in time

twilco replied:
When notifications are processed as a result of `postNotification`, the isolated tree is updated at the same time (and actually, before the notification goes out to ATs):

```cpp
void AXObjectCache::notificationPostTimerFired()
{
    ...truncated...

    auto notifications = std::exchange(m_notificationsToPost, { });

    ...truncated...

    updateIsolatedTree(notificationsToPost);

    for (const auto& note : notificationsToPost)
        postPlatformNotification(note.first, note.second);
}
```

So I think it should be OK, but let me know if I'm not understanding you correctly.

twilco committed:
AX: REGRESSION(286869@main): AXPropertyName::IsExpanded becomes stale for details and summary elements, causing accessibility/mac/details-summary.html to fail in ITM
https://bugs.webkit.org/show_bug.cgi?id=283578
rdar://140428627

Reviewed by Chris Fleizach.

Prior to https://commits.webkit.org/286869@main, we used to get away with not updating this property because every time
the details were expanded, we would throw away the entire details, summary, and other descendant objects and create new
ones (because the DOM implementation destroyed and recreated the DOM elements each time). After
https://commits.webkit.org/286869@main, the DOM implementation no longer destroys and recreates brand new DOM elements,
instead managing the visibility via CSS content-visibility. This a good change, since we now do significantly less work,
but need to make sure this property stays up-to-date. This commit does that.

* LayoutTests/accessibility/mac/details-summary-expected.txt:
* LayoutTests/accessibility/mac/details-summary.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onExpandedChanged):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::toggleOpen):

Canonical link: https://commits.webkit.org/287009@main
